pkg/experiment/distributor/placement/adaptive_placement/placement_manager.go

```diff
@@ -0,0 +1,210 @@
+package adaptive_placement
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/services"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/pyroscope/pkg/experiment/distributor/placement/adaptive_placement/adaptive_placementpb"
+	"github.com/grafana/pyroscope/pkg/iter"
+	"github.com/grafana/pyroscope/pkg/util"
+)
+
+// Manager maintains placement rules and distribution stats in the store.
+//
+// Manager implements services.Service interface for convenience, but it's
+// meant to be started and stopped explicitly via Start and Stop calls.
+//
+// If manager is being stopped while updating rules, an ongoing attempt is
+// not aborted: we're interested in finishing the operation so that the rules
+// reflect the most recent statistics. Another reason is that another instance
+// might be already running at the Stop call time.
+//
+// When just started, the manager may not have enough statistics to build
+// the rules: StatsConfidencePeriod should expire before the first update.
+// Note that ruler won't downscale datasets for a certain period of time
+// after the ruler is created regardless of the confidence period. Therefore,
+// it's generally safe to publish rules even with incomplete statistics;
+// however, this allows for delays in response to changes of the data flow.
+type Manager struct {
+	started   atomic.Bool
+	startedAt time.Time
+
+	service services.Service
+	logger  log.Logger
+	config  Config
+	limits  Limits
+	metrics *managerMetrics
+
+	store Store
+	stats *DistributionStats
+	ruler *Ruler
+}
+
+func NewManager(
+	logger log.Logger,
+	reg prometheus.Registerer,
+	config Config,
+	limits Limits,
+	store Store,
+) *Manager {
+	m := &Manager{
+		logger:  logger,
+		config:  config,
+		limits:  limits,
+		store:   store,
+		stats:   NewDistributionStats(config.StatsAggregationWindow),
+		metrics: newManagerMetrics(reg),
+	}
+	m.service = services.NewTimerService(
+		config.PlacementUpdateInterval,
+		m.starting,
+		m.updateRulesNoError,
+		m.stopping,
+	)
+	return m
+}
+
+func (m *Manager) Service() services.Service { return m.service }
+
+func (m *Manager) RecordStats(samples iter.Iterator[Sample]) { m.stats.RecordStats(samples) }
+
+func (m *Manager) Start() { m.started.Store(true) }
+func (m *Manager) Stop()  { m.started.Store(false) }
+
+func (m *Manager) starting(context.Context) error { return nil }
+func (m *Manager) stopping(error) error           { return nil }
+
+// The function is only needed to satisfy the services.OneIteration
+// signature: there's no case when the service stops on its own:
+// it's better to serve outdated rules than to not serve at all.
+func (m *Manager) updateRulesNoError(ctx context.Context) error {
+	util.Recover(func() { m.updateRules(ctx) })
+	return nil
+}
+
+func (m *Manager) updateRules(ctx context.Context) {
```

Comment:
I doubt Go benchmarks make sense here. What part is concerning you the most?

Unless we're talking about dozens of millions of entries, I have no concerns about performance and scalability. Nevertheless, to get millions of entries here, a malicious user would need to send millions of requests with unique service names within the rules retention period; by that time, it will be blocked because of the rate limit.

However, we will limit the number of services per tenant to prevent DDoS in distributors because of other reasons – mainly because of the impact on the metastore; and this will be done based on the placement rules.

I don't see how data isolation at the tenant level might be beneficial.

Reply:
BTW, the most expensive part here (`updateRules`) is sorting, which we don't need, strictly speaking. Therefore, If you insist, I'll remove it to save another ~0.25% of CPU time :)

<img width="1691" alt="image" src="https://github.com/user-attachments/assets/2b979231-d48b-462b-ac1f-7dcbc2fcdeb0">

I was more concerned with `RecordStats`, but in practice, it utilizes less than logging here
https://github.com/grafana/pyroscope/blob/596641bfb9ae75748a828c9eded29aae3190a864/pkg/experiment/metastore/metastore_state_add_block.go#L16-L32

<img width="1687" alt="image" src="https://github.com/user-attachments/assets/78010a6d-575b-4672-aaba-2ebc84d5bb98">